### PR TITLE
feat(fibers): add persistent EventCount waiters

### PR DESCRIPTION
**Pull request by glevkovich**
feat(fibers): add persistent EventCount waiters

Adds support for persistent waiters that remain linked in the
EventCount queue across multiple notifications, enabling
callback-based waiters to survive multiple wake/sleep cycles
without re-subscribing.

Why this is needed:
- Fiber-based waiters (await()) create new Waiter instances per
  iteration and work fine with auto-unlink behavior
- Callback-based waiters (IoEvent) reuse the same Waiter instance
  and would be orphaned if auto-unlinked on notification
- in dragonfly: V2 backpressure uses subscribe_persistent() once with io_event_
  waiter, then multiple io_event_.await() calls

Implementation:
- Added persistent_ flag to Waiter with is_persistent()/
  set_persistent() accessors
- Modified NotifyOne/NotifyAll to skip unlinking persistent waiters
- Added subscribe_persistent() API returning RAII SubKey guard
- SubKey destructor unconditionally resets persistent flag and
  unlinks on cleanup for safe waiter reuse
- Waiter move constructor now preserves persistent flag

Safety & Documentation:
- Added DCHECK in NotifyOne to prevent starvation when mixing
  persistent and one-shot waiters (fires in debug builds)
- Documented iterator safety in NotifyAll (no invalidation due to
  deferred fiber activation and pre-advanced iterators)
- Added callback re-entrancy deadlock warning in API docs
- Comprehensive class-level documentation explaining one-shot vs
  persistent subscription modes and usage guidelines

Testing:
- 7 new tests covering: multi-fire notifications, RAII cleanup,
  SubKey move semantics, waiter reuse, mixed queue scenarios,
  single waiter edge case, and starvation guard (death test)

Signed-off-by: Gil Levkovich <69595609+glevkovich@users.noreply.github.com>

## Summary by CodeRabbit

* **New Features**
  * Persistent subscriptions for event notifications so waiters can remain active across repeated notifications.

* **Improvements**
  * Notification dispatch updated to notify each waiter exactly once while preserving persistent subscribers and avoiding unnecessary removals.

* **Bug Fixes**
  * Added a debug check to detect potential starvation when persistent subscribers are queued ahead of others.

* **Tests**
  * New unit tests covering persistent subscription behavior, RAII unsubscription, move semantics, mixed queues, and starvation checks.

### CODE REVIEW

File: util/fibers/detail/wait_queue.cc [hunk 1]
```diff
@@ -37,16 +37,36 @@ bool WaitQueue::NotifyOne(FiberInterface* active) {
...
   Waiter* waiter = &wait_list_.front();
   DCHECK(waiter) << wait_list_.empty();
-  wait_list_.pop_front();
+
+  // If the waiter is one-shot (default), remove it from the list before
+  // notification. If it is persistent, leave it at the front so it can
+  // receive subsequent notifications until explicitly unlinked.
+  if (!waiter->is_persistent()) {
+    wait_list_.pop_front();
+  }
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/fibers/detail/wait_queue.cc:53]: fixed with DCHECK. we disallow notifyone with persistent waiter and multiple elements in queue

File: util/fibers/synchronization.h [hunk 2]
```diff
@@ -588,6 +608,18 @@ std::optional<EventCount::SubKey> EventCount::check_or_subscribe(Condition condi
...
+inline EventCount::SubKey EventCount::subscribe_persistent(detail::Waiter* w) noexcept {
+  DCHECK(!w->IsLinked());
+  w->set_persistent(true);
+  Key key = prepareWait();
+  std::unique_lock lk(lock_);
+
+  // No epoch check is needed because persistent waiters remain linked across all
+  // notifications, eliminating the risk of a "lost wakeup" during the subscription gap.
+  wait_queue_.Link(w);
+  return SubKey{std::move(key), w};
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/fibers/synchronization.h:626]: dropping. changedcomment to be more clear:

  // No epoch check needed here (unlike one-shot subscribe()). One-shot waiters need the
  // epoch check because they receive exactly one notification - missing the gap notification
  // means sleeping forever. Persistent waiters remain linked across ALL future notifications,
  // so even if a notification fires in the prepareWait()-to-Link() gap, the next notification
  // will reach this waiter. The caller is expected to check its predicate after subscribing.

File: util/fibers/detail/wait_queue.h [hunk 3]
```diff
@@ -46,6 +47,16 @@ class Waiter {
     return wait_hook.is_linked();
   }
 
+  // Persistent waiters are NOT unlinked by WaitQueue::NotifyAll();
+  // their callback is fired but they remain in the queue.
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/fibers/detail/wait_queue.h]: fixed

File: util/fibers/synchronization.h [hunk 4]
```diff
@@ -94,7 +110,11 @@ class EventCount {
   template <typename Condition>
   std::optional<SubKey> check_or_subscribe(Condition condition, detail::Waiter* w);
 
-  // Advanced API, most use-cases will requie await function.
+  // Subscribe a callback-based waiter persistently: notifyAll() fires its callback
+  // but does NOT unlink it. Returns a RAII guard that unlinks on destruction.
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/fibers/synchronization.h]: fixed

File: util/fibers/synchronization.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/fibers/synchronization.h:626]: REPEATING comment. addressed dropping.

File: util/fibers/detail/wait_queue.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/fibers/detail/wait_queue.cc:53]: Dismissing. We deliberately chose not to rotate persistent waiters for performance reasons.
We DCHECK for that.

[the remark this answers is not in the document]
Comment (glevkovich) [reply util/fibers/detail/wait_queue.cc:53]: Dismissing. We deliberately chose not to rotate persistent waiters for performance reasons.
added dcheck.

File: util/fibers/synchronization.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/fibers/synchronization.h]: fixed already. dropping

File: util/fibers/detail/wait_queue.cc [hunk 5]
```diff
@@ -37,16 +37,36 @@ bool WaitQueue::NotifyOne(FiberInterface* active) {
...
+  // until they are explicitly destroyed or unlinked via RAII.
+  while (it != wait_list_.end()) {
+    Waiter& waiter = *it;
+    if (waiter.is_persistent()) {
+      ++it;
+    } else {
+      it = wait_list_.erase(it);
+    }
+    waiter.Notify(active);
     notified = true;
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/fibers/detail/wait_queue.cc]: dropping. added documentation: 
```
// IMPORTANT: The callback must NOT call unsubscribe() or destroy SubKey, because
  // notification runs under EventCount's internal spinlock and re-entering it would
  // deadlock. The callback should only set a flag or perform a non-blocking signal.
```

File: util/fibers/synchronization.h [hunk 6]
```diff
@@ -8,6 +8,7 @@
 #include <condition_variable>  // for cv_status
 #include <optional>
 
+#include "base/logging.h"
```
Comment (romange) [inline util/fibers/synchronization.h]: do not add include `"base/logging.h"` in other header files.
use assert if needed.

File: util/fibers/synchronization.h [hunk 7]
```diff
@@ -53,20 +61,28 @@ class EventCount {
...
    public:
     SubKey(SubKey&& other) noexcept = default;
 
     ~SubKey() {
-      if (me_ != nullptr)
-        me_->unsubscribe(waiter);
+      if (me_ != nullptr) {
+        me_->unsubscribe(waiter_);
+        if (waiter_->is_persistent()) {
+          waiter_->set_persistent(false);
```
Comment (romange) [inline util/fibers/synchronization.h]: you can call `waiter_->set_persistent(false);` unconditionally.

Comment (glevkovich) [reply util/fibers/synchronization.h]: fixed

File: util/fibers/synchronization.h (hunk 6 above)
Comment (glevkovich) [reply util/fibers/synchronization.h]: fixed

File: util/fibers/detail/wait_queue.cc [hunk 8]
```diff
@@ -37,16 +37,36 @@ bool WaitQueue::NotifyOne(FiberInterface* active) {
 
   Waiter* waiter = &wait_list_.front();
   DCHECK(waiter) << wait_list_.empty();
-  wait_list_.pop_front();
+
+  // If the waiter is one-shot (default), remove it from the list before
```
Comment (romange) [inline util/fibers/detail/wait_queue.cc:48]: this one is problematic, as you starve other waiters. maybe pop_front and push_back?

Comment (glevkovich) [reply util/fibers/detail/wait_queue.cc:48]: We actually prohibit this case entirely now. I’m about to push a lot of changes once I finish my local testing and address the AI bot reviews. There is now a DCHECK that prevents using NotifyOne on a persistent waiter if it isn't the only one in the list, and I’ve added an explanation for why we don't support that case for now. Please wait for the next push.

File: util/fibers/synchronization.h [hunk 9]
```diff
@@ -588,6 +611,21 @@ std::optional<EventCount::SubKey> EventCount::check_or_subscribe(Condition condi
...
+  w->set_persistent(true);
+  Key key = prepareWait();
+  std::unique_lock lk(lock_);
+
+  // No epoch check needed here (unlike one-shot subscribe()). One-shot waiters need the
+  // epoch check because they receive exactly one notification - missing the gap notification
+  // means sleeping forever. Persistent waiters remain linked across ALL future notifications,
+  // so even if a notification fires in the prepareWait()-to-Link() gap, the next notification
+  // will reach this waiter. The caller is expected to check its predicate after subscribing.
+  wait_queue_.Link(w);
```
Comment (dranikpg) [inline util/fibers/synchronization.h]: depends on how you use it, missing a notification could be a problem in some cases

**Review by dranikpg — DISMISSED** [review]

File: util/fibers/synchronization.h (hunk 9 above)
Comment (glevkovich) [reply util/fibers/synchronization.h]: You're completely right that a notification in that gap is missed. However, because EventCount is a state-based primitive and not an event stream (in other words it protects state, it isn't a pub/sub event stream) , I intentionally left the epoch loop out.

If a notify() happens during the "gap", it means the condition  just became true.
Even though the waiter missed the notify() signal, the very next thing the code does is check the condition. It will see that the condition is true, and it will just keep running without ever going to sleep. No freeze happens.

Adding an epoch-retry loop here adds unnecessary atomic overhead just to protect against a caller misusing the API (i.e., subscribing and waiting without checking their condition first). 

I updated the comments to explicitly state this invariant.

**Review by dranikpg — APPROVED** [review]
